.. _check_fastq:

Valid FASTQ and Common Problems
-----------------------------------
The processing protocol permits the user to upload FASTQ files. The processing can only handle FASTQ files produced from Illumina HiSeq/MiSeq **only** and not just any FASTQ file generated by the end-user. Currently, we expect that for every sequence file there is an accompanying barcode file (1-to-1 relationship). Before uploading any FASTQ, we strongly encourage end-users to gzip their FASTQ files. This helps alleviate some of the storage issues that we have come across on the Web Server.

Common Problems
^^^^^^^^^^^^^^^^^
The most common problems with FASTQ files is determining which sequences file to use, along with failed uploading of the files through JumpLoader.

Expected 
************
The formatting of the files is not normally a problem, since they are raw files. The expected format for the amplicon read FASTQ should be as follows:

::

    @M10_68:1:1:28680:29475#0/1
    AACGAAAGGCAGTTTTGGAAGTAGGCGAATTAGGGTAACGCATATAGGATGCTAATACAACGTGAATGAAGTACTGCATCTATGTCACCAGCTTATTACAGCAGCTTGTCATACATGGCCGTACAGGAAACACACATCATAGCATCACACGA
    +
    BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB
    @M10_68:1:1:19607:29475#0/1
    GACATAAGGGTGGTTAGTATACCGGCAAGGACGGGGTTACTAGTGACGTCCTTCCCCGTATGCCGGGCAATAATGTTTATGTTGGTTTCATGGTTTGGTCTAACTTTACCGCTACTAAATGCTGCGGATTGGTTTCGCTGAATCAGATTATT
    +
    Z__c\JQ`cc[[_[bfff[[`Qbdge_YYOOHO^cF[FUb_VHMHV`T`dBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB

The expected format for the accompanying barcode FASTQ should be as follows:

::

    @M10_68:1:1:28680:29475#0/2
    ACTCACGGTATT
    +
    \_J\Sa^Y[ZYK
    @M10_68:1:1:19607:29475#0/2
    AGACTGAGTACT
    +
    PP\JJ\JQ`\RK

.. _check_fastq_sol1:

Problem 1 - Determining Sequence Files to Use
***********************************************
Most of the time, the sequencer will produce 2 amplicon read files (5' and 3' sequencing) along with a single barcode file. 

Solution
*************
When there are 3 files (e.g., s_1_1_sequence.fastq, s_1_2_sequence.fastq, s_1_3_sequence.fastq) produced we have always been taking the first 2 files produced (s_1_1_sequence.fastq and s_1_2_sequence.fastq) where the first file (s_1_1_sequence.fasta) is normally the 5' amplicon read file and the second file (s_1_2_sequence.fastq) is normally the barcode file. As a quick test, you can do an "ls" in the directory and you should notice that the second file is smaller than the first file. Essentially we do not do anything with the third file produced. A normal usage of that file is for performing taxonomic assignment and comparing against the results from the first file. For all of the studies coming from our lab or in collaboration with our lab, this method has always held true, but this does not mean it could change in the future. 

Problem 2 - Unable to Upload with JumpLoader
***********************************************
Since the amount of data being produced from sequencers, we have run into some bottlenecks uploading the FASTQ files through the website. As mentioned above, we strongly encourage end-users to gzip their FASTQ files prior to upload.

Solution
*************
In the case that users are unable to upload their FASTQ files via the website, we encourage them to send us a link to the files (or a location on a server that we have access to, such as compy). Once we know the location of the files, verify we know the exact lane the study corresponds to, since it is common for all lanes to be in the same folder. Once we know the file locations we can "scp" the files to the Web Server as follows:

:: 
    
    scp s_1_1_sequence.fastq [user-acct]@www.microbio.me:/home/wwwuser/
    
Next you should login to the Web Server, change the ownership on the FASTQ files and move them into the appropriate study location. 

::
    
    ### Login to your account
    ssh [user-acct]@www.microbio.me
    
    ### From your user-account, type the following:
    sudo chown wwwuser /home/wwwuser/s_1_1_sequence.fastq
    sudo chgrp wwwuser /home/wwwuser/s_1_1_sequence.fastq
    sudo su - wwwuser
    mv /home/wwwuser/s_1_1_sequence.fastq /home/wwwuser/user_data/studies/study_#/
    
Once the files are moved into place, we need to add the filepaths in the STUDY_FILES table under the QIIME_METADATA tablespace. For instance if the file above was part of study 0, then we would need to run the following SQL after logging into the QIIME_METADATA tablespace.

::

    INSERT INTO study_files (study_id, file_path, file_type) 
    VALUES (0, '/home/wwwuser/user_data/studies/study_0/s_1_1_sequence.fastq','FASTQ');
    COMMIT;
    
Once all the files have been moved into place you should verify that the sequence files appear under the Study Page on the website. You should also make sure that the RUN_PREFIX defined in the metadata corresponds to the sequence files you moved into place.